File: Core/src/burlton/core/code/util/XmlUtil.java
package burlton.core.code.util;

import burlton.core.code.obj.SuperHashMap;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import java.io.StringReader;
import java.io.StringWriter;
import java.util.Iterator;
import java.util.Map;

public class XmlUtil 
{
	//Cache one instance of the factories - I've seen from thread stacks that constructing it each time is really slow
	private static final DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
	private static final TransformerFactory tf = TransformerFactory.newInstance();
	
	public static Document factoryNewDocument()
	{
		try
		{
			DocumentBuilder builder = factory.newDocumentBuilder();
		
			return builder.newDocument();
		}
		catch (Throwable t)
		{
			Debug.stackTrace(t);
			return null;
		}
	}
	
	public static String getStringFromDocument(Document xmlDoc)
	{
		try
		{
			Transformer transformer = tf.newTransformer();
			transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
			StringWriter writer = new StringWriter();
			transformer.transform(new DOMSource(xmlDoc), new StreamResult(writer));

			StringBuffer sb = writer.getBuffer();
			String s = sb.toString().replaceAll("\n|\r", "");
			return s;
		}
		catch (Throwable t)
		{
			Debug.stackTrace(t);
			return "";
		}
	}
	
	public static Document getDocumentFromXmlString(String xmlStr)
	{
		try
		{
			DocumentBuilder builder = factory.newDocumentBuilder();
			builder.setErrorHandler(null);
			return builder.parse(new InputSource(new StringReader(xmlStr)));
		}
		catch (Throwable t)
		{
			return null;
		}
	}
	
	public static int getAttributeInt(Element rootElement, String attributeName)
	{	
		return getAttributeInt(rootElement, attributeName, 0);
	}
	
	public static int getAttributeInt(Element rootElement, String attributeName, int defaultValue)
	{
		String attribute = rootElement.getAttribute(attributeName);
		if (attribute.equals(""))
		{
			return defaultValue;
		}
		
		int ret = -1;
		try
		{
			ret = Integer.parseInt(attribute);
		}
		catch (NumberFormatException nfe)
		{
			Debug.append("Caught nfe trying to parse attribute " + attributeName + " with value " + attribute);
		}
		
		return ret;
	}
	
	public static double getAttributeDouble(Element rootElement, String attributeName)
	{
		String attribute = rootElement.getAttribute(attributeName);
		if (attribute.equals(""))
		{
			return 0;
		}
		
		return Double.parseDouble(attribute);
	}
	
	public static void setAttributeBoolean(Element rootElement, String attributeName, boolean bool)
	{
		if (bool)
		{
			rootElement.setAttribute(attributeName, "" + bool);
		}
	}

	public static Document factorySimpleMessage(String username, String rootName)
	{
		Document message = factoryNewDocument();
		
		Element rootElement = message.createElement(rootName);
		rootElement.setAttribute("Username", username);
		
		message.appendChild(rootElement);
		return message;
	}
	
	public static SuperHashMap<Integer, Integer> readIntegerHashMap(Element rootElement, String tagName, String keyTag, String valueTag)
	{
		SuperHashMap<Integer, Integer> hm = new SuperHashMap<>();
		
		NodeList children = rootElement.getElementsByTagName(tagName);
		int size = children.getLength();
		for (int i=0; i<size; i++)
		{
			Element child = (Element)children.item(i);
			int key = XmlUtil.getAttributeInt(child, keyTag);
			int value = XmlUtil.getAttributeInt(child, valueTag);
			
			hm.put(key, value);
		}
		
		return hm;
	}
	
	@SuppressWarnings("rawtypes")
	public static void writeHashMap(SuperHashMap hm, Document xmlDoc, Element rootElement, 
	  String tagName, String keyTag, String valueTag)
	{
		Iterator<Map.Entry> it = hm.entrySet().iterator();
		for (; it.hasNext(); )
		{
			Map.Entry entry = it.next();
			
			Element child = xmlDoc.createElement(tagName);
			child.setAttribute(keyTag, "" + entry.getKey());
			child.setAttribute(valueTag, "" + entry.getValue());
			rootElement.appendChild(child);
		}
	}
}

File: Core/src/burlton/core/code/util/XmlUtil.kt
package burlton.core.code.util

import org.w3c.dom.Document
import org.w3c.dom.Element
import org.xml.sax.InputSource
import java.io.StringReader
import java.io.StringWriter
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.OutputKeys
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

//Cache one instance of the factories - I've seen from thread stacks that constructing it each time is really slow
private val factory = DocumentBuilderFactory.newInstance()
private val tf = TransformerFactory.newInstance()

object XmlUtil
{
    fun factoryNewDocument(): Document
    {
        val builder = factory.newDocumentBuilder()
        return builder.newDocument()
    }

    fun getStringFromDocument(xmlDoc: Document): String
    {
        try
        {
            val transformer = tf.newTransformer()
            transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes")
            val writer = StringWriter()
            transformer.transform(DOMSource(xmlDoc), StreamResult(writer))

            val sb = writer.buffer
            return sb.toString().replace("\n|\r".toRegex(), "")
        }
        catch (t: Throwable)
        {
            Debug.stackTrace(t)
            return ""
        }

    }

    fun getDocumentFromXmlString(xmlStr: String) =
        try
        {
            val builder = factory.newDocumentBuilder()
            builder.setErrorHandler(null)
            builder.parse(InputSource(StringReader(xmlStr)))
        }
        catch (t: Throwable)
        {
            null
        }
}

/**
 * Extension fns
 */
fun Element.getAttributeInt(attributeName: String, defaultValue: Int = 0): Int
{
    val attribute = getAttribute(attributeName)
    return if (attribute == "") defaultValue else attribute.toInt()
}

fun Element.getAttributeDouble(attributeName: String): Double
{
    val attribute = getAttribute(attributeName)
    return if (attribute == "") 0.0 else attribute.toDouble()
}

fun Element.readIntegerHashMap(tagName: String, keyTag: String, valueTag: String): MutableMap<Int, Int>
{
    val hm = mutableMapOf<Int, Int>()

    val children = getElementsByTagName(tagName)
    val size = children.length
    for (i in 0 until size)
    {
        val child = children.item(i) as Element
        val key = child.getAttributeInt(keyTag)
        val value = child.getAttributeInt(valueTag)

        hm[key] = value
    }

    return hm
}

fun Element.writeHashMap(hm: Map<*, *>, tagName: String, keyTag: String, valueTag: String)
{
    hm.forEach { key, value ->
        val child = ownerDocument.createElement(tagName)
        child.setAttribute(keyTag, "$key")
        child.setAttribute(valueTag, "$value")
        appendChild(child)
    }
}

fun Element.setAttribute(key: String, value: Any) = setAttribute(key, "$value")

fun Document.createRootElement(name: String) = createElement(name).also { appendChild(it) }
